trace-server/org.eclipse.tracecompass.incubator.trace.server.jersey.rest.core/src/org/eclipse/tracecompass/incubator/internal/trace/server/jersey/rest/core/services/Bookmark.java

```diff
@@ -0,0 +1,136 @@
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * Bookmark model for TSP
+ *
+ * @author Kaveh Shahedi
+ * @since 10.1
+ */
+public class Bookmark implements Serializable {
+    private static final long serialVersionUID = -3626414315455912960L;
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    private final UUID fUUID;
+    private final String fName;
+    private final String fExperimentId;
+    private final long fStart;
+    private final long fEnd;
+    private final JsonNode fPayload;
+
+    /**
+     * {@link JsonCreator} Constructor for final fields
+     *
+     * @param uuid
+     *            the stub's UUID
+     * @param name
+     *            bookmark name
+     * @param experimentId
+     *            experiment id
+     * @param start
+     *            start time
+     * @param end
+     *            end time
+     * @param payload
+     *            additional JSON data associated with the bookmark (optional)
+     */
+    @JsonCreator
+    public Bookmark(
+            @JsonProperty("UUID") UUID uuid,
+            @JsonProperty("name") String name,
+            @JsonProperty("experimentId") String experimentId,
+            @JsonProperty("start") long start,
+            @JsonProperty("end") long end,
+            @JsonProperty(value = "payload", required = false) JsonNode payload) {
+        fUUID = uuid;
+        fName = name;
+        fExperimentId = experimentId;
+        fStart = start;
+        fEnd = end;
+        fPayload = (payload != null) ? payload : MAPPER.createObjectNode();
+    }
+
+    /**
+     * Constructor without payload
+     *
+     * @param uuid
+     *            the stub's UUID
+     * @param name
+     *            bookmark name
+     * @param experimentId
+     *            experiment id
+     * @param start
+     *            start time
+     * @param end
+     *            end time
+     */
+    public Bookmark(UUID uuid, String name, String experimentId, long start, long end) {
+        this(uuid, name, experimentId, start, end, MAPPER.createObjectNode());
+    }
+
+    /**
+     * Get the UUID
+     *
+     * @return the UUID
+     */
+    public UUID getUUID() {
+        return fUUID;
+    }
+
+    /**
+     * Get the bookmark name
+     *
+     * @return the bookmark name
+     */
+    public String getName() {
+        return fName;
+    }
+
+    /**
+     * Get the experiment id
+     *
+     * @return the experiment id
+     */
+    public String getExperimentId() {
+        return fExperimentId;
+    }
+
+    /**
+     * Get the start time
+     *
+     * @return the start time
+     */
+    public long getStart() {
+        return fStart;
+    }
+
+    /**
+     * Get the end time
+     *
+     * @return the end time
+     */
+    public long getEnd() {
+        return fEnd;
+    }
+
+    /**
+     * Get the payload
+     *
+     * @return the JSON payload, empty JSON object if no payload was set
```

Comment:
what is expected as payload? As a FE designer I would not know what to put there.

Reply:
This comes back to the requirements of Trace-Server/TraceCompass. This field is supposed to contain some metadata of other types of information that might be required to be saved in a bookmark. Since the fields of bookmark were not clear to me, I did put this extra field to be more flexible.

Reply:
The point of the TSP is that the clients (front-end) can make sense of the data. If a server is returning arbitrary data, then a generic client implementation won't know how to input the data or display the data. Otherwise, we need custom client implementation for each custom bookmark data. 

In the application, we can have multiple types of bookmarks, which have common fields and bookmark type specific fields. What you're introducing I would call global bookmark, which has a start, end (optional), name, description. It may have a colour associated with (OutputElementStyle). This is similar to lost event markers. 

In the future, and it's not in scope for the initial bookmark support, we might create bookmarks on specific views, series, rows etc. which will have additional information, that needs to be stored in the bookmark.

What is in scope, however, is that the data structures that we define know can be augmented to support other use cases.